### PR TITLE
Add simple desktop notification for logout

### PR DESCRIPTION
**Pull request by MarcoPolo**
### Context

We'll want desktop notifications on the app. It'll work through the desktop app and be sent from the service.

@chrisnojima and I talked yesterday and think it might make more sense to handle all the notifications when we build the shim app. However we'd still need to figure out the basics of notifications. This does just that.

### Changes to Engine
This is the first rpc request where the service can, at any point, send a message to the client. I've tentatively named it `generalIncomingRpc` (but we can change it). 

Added: 
* bind to a `generalIncomingRpc` with a method name.
* Unbind ^

We'll also want the ability to enable notifications as soon as we connect to the transport, so now there is a new `engine.listenOnConnect` that will let call a callback when the engine has connected.

### Notification type

I figured the easiest thing we could do is reuse chormiums existing [desktop notification](https://developer.mozilla.org/en-US/docs/Web/API/notification). It uses the native notification on mac and works the same on linux/windows. I tested it with (awesomewm and kde) on ~~windows~~ linux.

@keybase/react-hackers

### CODE REVIEW

**Comment by cjb** [comment]
> I tested it with (awesomewm and kde) on windows.

Wait am I reading that right?  :)

File: desktop/renderer/index.desktop.js [hunk 1]
```diff
@@ -9,6 +9,7 @@ import configureStore from '../../react-native/react/store/configure-store'
 import Nav from '../../react-native/react/nav'
 import { DevTools, DebugPanel, LogMonitor } from 'redux-devtools/lib/react'
 import injectTapEventPlugin from 'react-tap-event-plugin'
+import { enableNotifications, bindNotifications, unbindNotifications} from '../../react-native/react/native/notifications'
```
Comment (chrisnojima) [inline desktop/renderer/index.desktop.js]: do we need to reach into react-native to get this? doesn't desktop have react directly?

File: react-native/react/engine/index.js [hunk 2]
```diff
@@ -35,6 +37,25 @@ class Engine {
...
+
+    // A list of functions to call when we connect
+    this.onConnectFns = []
+  }
+
+  onConnect () {
+    this.onConnectFns.forEach(f => f())
+  }
+
+  listenOnConnect (f) {
```
Comment (chrisnojima) [inline react-native/react/engine/index.js]: do we need to worry about reconnect and having to bookkeep this?

File: react-native/react/engine/index.js [hunk 3]
```diff
@@ -57,10 +78,29 @@ class Engine {
     )
   }
 
-  _rpcWrite (data) {
+  listenGeneralIncomingRpc (method, listener) {
+    if (this.generalListeners[method] == null) {
```
Comment (chrisnojima) [inline react-native/react/engine/index.js]: prefer !this.generalListeners[method] or ===

File: react-native/react/engine/index.js [hunk 4]
```diff
@@ -110,6 +150,8 @@ class Engine {
       }
 
       callMap[method](param, wrappedResponse)
+    } else if (sessionID == null && this.generalListeners[method]) {
```
Comment (chrisnojima) [inline react-native/react/engine/index.js]: prefer !sessionID

File: react-native/react/native/notifications.desktop.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
+'use strict'
+/* @flow */
+
+import engine from '../engine'
+
+const param = {
```
Comment (chrisnojima) [inline react-native/react/native/notifications.desktop.js]: lets move this into enableNotifications

File: react-native/react/native/notifications.desktop.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+export function enableNotifications () {
+  console.log('setting up notification')
+  engine.listenOnConnect(() => {
+    engine.rpc('notifyCtl.toggleNotifications', param, {}, (error, response) => {
+      if (error != null) {
+        console.error('error in toggling notifications: ', error)
+      }
+      console.log('Enabled Notifications')
```
Comment (chrisnojima) [inline react-native/react/native/notifications.desktop.js]: should this be an else?

File: react-native/react/native/notifications.desktop.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    })
+  })
+}
+
+export function bindNotifications () {
```
Comment (chrisnojima) [inline react-native/react/native/notifications.desktop.js]: maybe we should call a helper that'll do this automatically so we don't need super similar code here and in the unbind?

**Comment by Sidnicious** [comment]
:shipit:

File: react-native/react/native/notifications.desktop.js (hunk 7 above)
Comment (Sidnicious) [inline react-native/react/native/notifications.desktop.js]: What if `listenGeneralIncomingRpc` returned a function that you could call (with no arguments) to unbind it?

**Pull request merged**